core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -148,35 +145,41 @@ protected boolean resolveIndex() {
         return location;
     }
 
-    private Translog.Location index(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) {
+    private Translog.Location index(IndexMetaData metaData, BulkShardRequest request, IndexShard indexShard, long[] preVersions, VersionType[] preVersionTypes, Translog.Location location, int requestIndex, BulkItemRequest item) throws Exception {
         IndexRequest indexRequest = (IndexRequest) item.request();
         preVersions[requestIndex] = indexRequest.version();
         preVersionTypes[requestIndex] = indexRequest.versionType();
         try {
             WriteResult<IndexResponse> result = shardIndexOperation(request, indexRequest, metaData, indexShard, true);
-            location = locationToSync(location, result.getLocation());
-            // add the response
-            IndexResponse indexResponse = result.getResponse();
-            setResponse(item, new BulkItemResponse(item.id(), indexRequest.opType().lowercase(), indexResponse));
+            if (result.success()) {
+                location = locationToSync(location, result.getLocation());
+                // add the response
+                IndexResponse indexResponse = result.getResponse();
+                setResponse(item, new BulkItemResponse(item.id(), indexRequest.opType().lowercase(), indexResponse));
+            } else {
+                logFailure(result.getFailure(), "index", request.shardId(), indexRequest);
+                // if its a conflict failure, and we already executed the request on a primary (and we execute it
+                // again, due to primary relocation and only processing up to N bulk items when the shard gets closed)
+                // then just use the response we got from the successful execution
+                if (item.getPrimaryResponse() != null && isConflictException(result.getFailure())) {
+                    setResponse(item, item.getPrimaryResponse());
+                } else {
+                    setResponse(item, new BulkItemResponse(item.id(), indexRequest.opType().lowercase(),
+                        new BulkItemResponse.Failure(request.index(), indexRequest.type(), indexRequest.id(), result.getFailure())));
+                }
+            }
         } catch (Exception e) {
```

Comment:
I wonder if this fits in the new philosophy - all "valid" item level failures should not be thrown but rather be channelled into result.getFailure(), so everything that comes in here should not be shoved into  an item level failure, but rather bubbled up. Since we want to have special logic in the retryPrimaryException , why not just catch ElasticsearchException? we can also addes asserts to retryPrimaryException to make sure it only returns true if it's paramater is  subclass of ElasticsearchException?